nitro-protocol/contracts/LedgerFinancingApp.sol

```diff
@@ -0,0 +1,151 @@
+pragma solidity 0.8.17;
+pragma experimental ABIEncoderV2;
+
+import './interfaces/IForceMoveApp.sol';
+import './libraries/NitroUtils.sol';
+import {ExitFormat as Outcome} from '@statechannels/exit-format/contracts/ExitFormat.sol';
+
+// LedgerFinancingApp is a ForceMoveApp that allows a intermediary to earn interest
+// on a deposit. It fuctions as a ConsensusApp with the following additional rule:
+//   - the intermediary can unilaterally transition from state n to state n+1,
+//     forcing calculated interest into the intermediary's Outcome allocation
+contract LedgerFinancingApp is IForceMoveApp {
+    struct Funds {
+        address[] asset;
+        uint256[] amount;
+    }
+
+    struct InterestAppData {
+        // a per-day simple interest rate (daily percentage yield), as a fraction.
+        // ie, 1% per day is represented with dpyNum = 1 and dpyDen = 100.
+        //
+        // Lower numbers (eg, fraction in simplest terms) produce least risk of overflow.
+        uint128 dpyNum;
+        uint128 dpyDen;
+        // the block number of the latest principal adjustment
+        uint256 blocknumber;
+        // the current principal. Decreases as the serviceProvider earns via the channel.
+        Funds principal;
```

Comment:
I had to look it up, but this seems like a good term to use https://www.investopedia.com/terms/p/principal.asp#toc-what-is-principal

Reply:
Should the `principal` ever change? I think we discussed this on notion... ideally application-specific constants could be stored in the IForceMoveApp itself (and passed as constructor params).

Reply:
OK I think I see how it works. 

**ON chain** -- there is only ever a _single_ transition allowed (from a ✔️ ✅  state to a ✅ state)
**OFF chain** -- the parties can agree to change `principal, blocknumber and Outcome` (according to the rules set out in this contract) by unanimous consensus (supporting a ✔️ ✅  state). 

In practice, is there any point in doing this off chain mechanism other than as a part of defunding the ledger channel? I suppose if `dpyDen` changed, they would want to "reset the counting" in this way.

Reply:
The checks here are signatures from each participant (green: serviceProvider, white: intermediary), yes?

re: off chain principal adjustment

As see it, the principal should diminish based on SP earnings in the channel. For Alice, Bob, Irene, channels `AI`, `IB`, and `XY_X := the amount in XY allocated to X`, balances change like this:

| time | IB_I | AI_I |
|------|------|------|
|   0  |  10  |   0  |
|   1  |   9  |   1  |

Here, Irene deposited 10 coins to onboard Bob to the network. Bob expects to make a profit, so he's agreed to pay a yield on the deposit.

But after `t = 1`, Irene's commitment in the channel is no longer 10. Her claimable assets have sloshed around in the network. I *think* it no longer makes good sense for Bob to be paying interest on the 1 coin belonging to Irene in `AI`. So we adjust `appData.principal` of `IB` down to nine.

From Irene's perspective, she still has ten coins "locked-in" to the network. Nine of them are still generating yield, one isn't. But the one that isn't is now explicitly "her problem" rather than "Bob's problem".